File: backend/app/modules/devices/schemas/topology.py
from app.common.schemas.base import ORMModel


class TopologyNode(ORMModel):
    id: int
    label: str
    device_type: str | None = None
    vendor: str | None = None
    model: str | None = None
    ip_address: str | None = None
    status: str
    criticality: str | None = None
    data_classification: str | None = None
    site_id: int | None = None
    responsible_user_id: int | None = None
    responsible_user_name: str | None = None
    integration_id: int | None = None


class TopologyEdge(ORMModel):
    id: int
    source: int
    target: int
    link_type: str
    port_source: str | None = None
    port_target: str | None = None
    vlan_id: int | None = None
    bandwidth_mbps: int | None = None


class TopologyCluster(ORMModel):
    label: str
    site_id: int | None = None
    node_ids: list[int]


class TopologyResponse(ORMModel):
    client_id: int
    client_name: str | None = None
    nodes: list[TopologyNode]
    edges: list[TopologyEdge]
    clusters: list[TopologyCluster]

File: frontend/src/lib/api/topology.ts
import { request, buildQuery } from "./client";

export type LinkType = "ethernet" | "wifi" | "wan" | "vpn" | "mpls" | "trunk" | "cloud_peering";

export type TopologyNode = {
  id: number;
  label: string;
  device_type?: string | null;
  vendor?: string | null;
  model?: string | null;
  ip_address?: string | null;
  status: string;
  criticality?: string | null;
  data_classification?: string | null;
  site_id?: number | null;
  responsible_user_id?: number | null;
  responsible_user_name?: string | null;
  integration_id?: number | null;
};

export type TopologyEdge = {
  id: number;
  source: number;
  target: number;
  link_type: LinkType;
  port_source?: string | null;
  port_target?: string | null;
  vlan_id?: number | null;
  bandwidth_mbps?: number | null;
};

export type TopologyCluster = {
  label: string;
  site_id?: number | null;
  node_ids: number[];
};

export type TopologyResponse = {
  client_id: number;
  client_name?: string | null;
  nodes: TopologyNode[];
  edges: TopologyEdge[];
  clusters: TopologyCluster[];
};

export type DeviceConnection = {
  id: number;
  source_device_id: number;
  target_device_id: number;
  link_type: LinkType;
  port_source?: string | null;
  port_target?: string | null;
  vlan_id?: number | null;
  bandwidth_mbps?: number | null;
  notes?: string | null;
  created_at?: string | null;
};

export type DeviceConnectionCreate = Omit<DeviceConnection, "id" | "created_at">;

export async function getTopology(clientId: number): Promise<TopologyResponse> {
  return request<TopologyResponse>(`/devices/topology${buildQuery({ client_id: clientId })}`);
}

export async function getConnections(clientId: number): Promise<DeviceConnection[]> {
  return request<DeviceConnection[]>(`/device-connections/${buildQuery({ client_id: clientId })}`);
}

export async function createConnection(payload: DeviceConnectionCreate): Promise<DeviceConnection> {
  return request<DeviceConnection>("/device-connections/", { method: "POST", body: JSON.stringify(payload) });
}

export async function deleteConnection(connectionId: number): Promise<{ message: string }> {
  return request<{ message: string }>(`/device-connections/${connectionId}`, { method: "DELETE" });
}
